# Projet ML: Estimation du coût d'un bien immobilier

## Importation et observation des données

In [221]:
#Importation des bibliothèques utiles et chargement du jeu Sberbank
import pandas as pd
import numpy as np
# read the data
df = pd.read_csv('sberbank.csv')
print (type(df))

<class 'pandas.core.frame.DataFrame'>


Now we are going th have a general view to explore our dataset

In [222]:
print(df.shape)
df

(30471, 292)


id   timestamp  full_sq  life_sq  floor  max_floor  material  \
0          1  2011-08-20       43     27.0    4.0        NaN       NaN   
1          2  2011-08-23       34     19.0    3.0        NaN       NaN   
2          3  2011-08-27       43     29.0    2.0        NaN       NaN   
3          4  2011-09-01       89     50.0    9.0        NaN       NaN   
4          5  2011-09-05       77     77.0    4.0        NaN       NaN   
...      ...         ...      ...      ...    ...        ...       ...   
30466  30469  2015-06-30       44     27.0    7.0        9.0       1.0   
30467  30470  2015-06-30       86     59.0    3.0        9.0       2.0   
30468  30471  2015-06-30       45      NaN   10.0       20.0       1.0   
30469  30472  2015-06-30       64     32.0    5.0       15.0       1.0   
30470  30473  2015-06-30       43     28.0    1.0        9.0       1.0   

       build_year  num_room  kitch_sq  ...  cafe_count_5000_price_2500  \
0             NaN       NaN       NaN  ...                           9   
1             NaN       NaN       NaN  ...                          15   
2             NaN       NaN       NaN  ...                          10   
3             NaN       NaN       NaN  ...                          11   
4             NaN       NaN       NaN  ...                         319   
...           ...       ...       ...  ...                         ...   
30466      1975.0       2.0       6.0  ...                          15   
30467      1935.0       4.0      10.0  ...                         313   
30468         NaN       1.0       1.0  ...                           1   
30469      2003.0       2.0      11.0  ...                          22   
30470      1968.0       2.0       6.0  ...                           5   

      cafe_count_5000_price_4000 cafe_count_5000_price_high  \
0                              4                          0   
1                              3                          0   
2                              3                          0   
3                              2                          1   
4                            108                         17   
...                          ...                        ...   
30466                          5                          0   
30467                        128                         24   
30468                          1                          0   
30469                          1                          1   
30470                          2                          0   

       big_church_count_5000  church_count_5000  mosque_count_5000  \
0                         13                 22                  1   
1                         15                 29                  1   
2                         11                 27                  0   
3                          4                  4                  0   
4                        135                236                  2   
...                      ...                ...                ...   
30466                     15                 26                  1   
30467                     98                182                  1   
30468                      2                 12                  0   
30469                      6                 31                  1   
30470                      7                 16                  0   

       leisure_count_5000  sport_count_5000  market_count_5000  price_doc  
0                       0                52                  4    5850000  
1                      10                66                 14    6000000  
2                       4                67                 10    5700000  
3                       0                26                  3   13100000  
4                      91               195                 14   16331452  
...                   ...               ...                ...        ...  
30466                   2                84                  6    7400000  
30467                  82 

-we have 30471 row and 292 columns in our dataset as well we can see the first and the last 5 rows in our dataset  
-the most interesting column is the last one price_doc which describe the price of the building  
-when viewing the table we can see that the data have many problems which can lead to conficts in our model for example we have a lot of missing values as well we have duplicated column id (one include in our dataset and the other is automatically generated by our code) thus we need to clean the data before excuting our model 

In [223]:
df.describe()

id       full_sq       life_sq         floor     max_floor  \
count  30471.000000  30471.000000  24088.000000  30304.000000  20899.000000   
mean   15237.917397     54.214269     34.403271      7.670803     12.558974   
std     8796.501536     38.031487     52.285733      5.319989      6.756550   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%     7620.500000     38.000000     20.000000      3.000000      9.000000   
50%    15238.000000     49.000000     30.000000      6.500000     12.000000   
75%    22855.500000     63.000000     43.000000     11.000000     17.000000   
max    30473.000000   5326.000000   7478.000000     77.000000    117.000000   

           material    build_year      num_room      kitch_sq         state  \
count  20899.000000  1.686600e+04  20899.000000  20899.000000  16912.000000   
mean       1.827121  3.068057e+03      1.909804      6.399301      2.107025   
std        1.481154  1.543878e+05      0.851805     28.265979      0.880148   
min        1.000000  0.000000e+00      0.000000      0.000000      1.000000   
25%        1.000000  1.967000e+03      1.000000      1.000000      1.000000   
50%        1.000000  1.979000e+03      2.000000      6.000000      2.000000   
75%        2.000000  2.005000e+03      2.000000      9.000000      3.000000   
max        6.000000  2.005201e+07     19.000000   2014.000000     33.000000   

       ...  cafe_count_5000_price_2500  cafe_count_5000_price_4000  \
count  ...                30471.000000                30471.000000   
mean   ...                   32.058318                   10.783860   
std    ...                   73.465611                   28.385679   
min    ...                    0.000000                    0.000000   
25%    ...                    2.000000                    1.000000   
50%    ...                    8.000000                    2.000000   
75%    ...                   21.000000                    5.000000   
max    ...                  377.000000                  147.000000   

       cafe_count_5000_price_high  big_church_count_5000  church_count_5000  \
count                30471.000000           30471.000000       30471.000000   
mean                     1.771783              15.045552          30.251518   
std                      5.418807              29.118668          47.347938   
min                      0.000000               0.000000           0.000000   
25%                      0.000000               2.000000           9.000000   
50%                      0.000000               7.000000          16.000000   
75%                      1.000000              12.000000          28.000000   
max                     30.000000             151.000000         250.000000   

       mosque_count_5000  leisure_count_5000  sport_count_5000  \
count       30471.000000        30471.000000      30471.000000   
mean            0.442421            8.648814         52.796593   
std             0.609269           20.580741         46.292660   
min             0.000000            0.000000          0.000000   
25%             0.000000            0.000000         11.000000   
50%             0.000000            2.000000         48.000000   
75%             1.000000            7.000000         76.000000   
max             2.000000          106.000000        218.000000   

       market_count_5000     price_doc  
count       30471.000000  3.047100e+04  
mean            5.987070  7.123035e+06  
std             4.889219  4.780111e+06  
min             0.000000  1.000000e+05  
25%             1.000000  4.740002e+06  
50%             5.000000  6.274411e+06  
75%            10.000000  8.300000e+06  
max            21.000000  1.111111e+08  

[8 rows x 276 columns]

here we are exploring the data representation in our dataset via the describe methow which show count,mean,std,min,(25%,50%,75%),max
we can see that we have here 276 columns and not 292 the reason for that is we 16 not numeric column on which we can't execute describe method

In [224]:
df.columns

Index(['id', 'timestamp', 'full_sq', 'life_sq', 'floor', 'max_floor',
       'material', 'build_year', 'num_room', 'kitch_sq',
       ...
       'cafe_count_5000_price_2500', 'cafe_count_5000_price_4000',
       'cafe_count_5000_price_high', 'big_church_count_5000',
       'church_count_5000', 'mosque_count_5000', 'leisure_count_5000',
       'sport_count_5000', 'market_count_5000', 'price_doc'],
      dtype='object', length=292)

## Traitement des données manquante

In this part we are going to get rid of null values, to do this we have two different methods :  
-The first method: we can do this by remplacing the null values with eihter the most frequante value or the median or with the mean value of the column; there's no optimal choise between them we can decide which value to be used based on the situation  
-The second method: we have to drop the column or the row with a lot of misssing value since remplacing them can lead to misleading the model results  

First of all let's see how many data are missing and in which column are they and their type

In [225]:
count=0
for col in df.columns:
    column_missing_values = np.mean(df[col].isnull())
    if column_missing_values>0 :
        print('{} - {} - {}%'.format(col,df[col].dtypes, round(column_missing_values*100)))
        count=count+1
print('{} columns with missing values'.format(count))

life_sq - float64 - 21%
floor - float64 - 1%
max_floor - float64 - 31%
material - float64 - 31%
build_year - float64 - 45%
num_room - float64 - 31%
kitch_sq - float64 - 31%
state - float64 - 44%
preschool_quota - float64 - 22%
school_quota - float64 - 22%
hospital_beds_raion - float64 - 47%
raion_build_count_with_material_info - float64 - 16%
build_count_block - float64 - 16%
build_count_wood - float64 - 16%
build_count_frame - float64 - 16%
build_count_brick - float64 - 16%
build_count_monolith - float64 - 16%
build_count_panel - float64 - 16%
build_count_foam - float64 - 16%
build_count_slag - float64 - 16%
build_count_mix - float64 - 16%
raion_build_count_with_builddate_info - float64 - 16%
build_count_before_1920 - float64 - 16%
build_count_1921-1945 - float64 - 16%
build_count_1946-1970 - float64 - 16%
build_count_1971-1995 - float64 - 16%
build_count_after_1995 - float64 - 16%
metro_min_walk - float64 - 0%
metro_km_walk - float64 - 0%
railroad_station_walk_km - float64 - 0%
railr

As we can see here there's a lot of data missing it variates between 0% approximatally and 47%
but before we solve this problem let's see their distribution in the rows

In [226]:
count=0
for row in df.iloc:
    rows_missing_values = np.mean(row.isnull())
    if rows_missing_values>0 :
        print('{} - {}%'.format(row[0], round(rows_missing_values*100)))
        count=count+1
print('{} row with missing values'.format(count))

1 - 3%
2 - 2%
3 - 2%
4 - 2%
5 - 2%
6 - 2%
7 - 2%
8 - 2%
9 - 2%
10 - 2%
11 - 2%
12 - 4%
13 - 2%
14 - 2%
15 - 2%
16 - 2%
17 - 3%
18 - 2%
19 - 2%
20 - 2%
21 - 6%
22 - 2%
23 - 2%
24 - 2%
25 - 2%
26 - 3%
27 - 2%
28 - 2%
29 - 3%
30 - 2%
31 - 2%
32 - 3%
33 - 2%
34 - 2%
35 - 3%
36 - 3%
37 - 2%
38 - 2%
39 - 2%
40 - 3%
41 - 2%
42 - 3%
43 - 3%
44 - 2%
45 - 2%
46 - 2%
47 - 2%
48 - 3%
49 - 2%
50 - 2%
51 - 2%
52 - 2%
53 - 2%
54 - 2%
55 - 2%
56 - 2%
57 - 3%
58 - 3%
59 - 2%
60 - 2%
61 - 2%
62 - 2%
63 - 2%
64 - 2%
65 - 2%
66 - 3%
67 - 2%
68 - 2%
69 - 2%
70 - 2%
71 - 2%
72 - 4%
73 - 2%
74 - 2%
75 - 2%
76 - 3%
77 - 2%
78 - 3%
79 - 3%
80 - 2%
81 - 2%
82 - 3%
83 - 3%
84 - 2%
85 - 2%
86 - 2%
87 - 2%
88 - 2%
89 - 2%
90 - 2%
91 - 2%
92 - 2%
93 - 3%
94 - 2%
95 - 2%
96 - 2%
97 - 3%
98 - 2%
99 - 3%
100 - 2%
101 - 2%
102 - 2%
103 - 2%
104 - 2%
105 - 2%
106 - 2%
107 - 3%
108 - 2%
109 - 2%
110 - 2%
111 - 2%
112 - 2%
113 - 4%
114 - 2%
115 - 2%
116 - 2%
117 - 3%
118 - 2%
119 - 2%
120 - 2%
121 - 2%
122 - 2%
123 - 2%
1

927 - 2%
928 - 4%
929 - 2%
930 - 2%
931 - 3%
932 - 2%
933 - 3%
934 - 2%
935 - 5%
936 - 2%
937 - 2%
938 - 3%
939 - 3%
940 - 3%
941 - 2%
942 - 2%
943 - 2%
944 - 2%
945 - 3%
946 - 2%
947 - 2%
948 - 2%
949 - 2%
950 - 4%
951 - 3%
952 - 2%
953 - 2%
954 - 2%
955 - 2%
956 - 2%
957 - 2%
958 - 4%
959 - 2%
960 - 2%
961 - 2%
962 - 2%
963 - 3%
964 - 2%
965 - 2%
966 - 2%
967 - 3%
968 - 2%
969 - 2%
970 - 2%
971 - 2%
972 - 2%
973 - 2%
974 - 2%
975 - 2%
976 - 2%
977 - 2%
978 - 2%
979 - 2%
980 - 3%
981 - 2%
982 - 2%
983 - 2%
984 - 2%
985 - 2%
986 - 3%
987 - 3%
988 - 2%
989 - 2%
990 - 2%
991 - 3%
992 - 2%
993 - 2%
994 - 2%
995 - 2%
996 - 2%
997 - 2%
998 - 2%
999 - 2%
1000 - 2%
1001 - 3%
1002 - 3%
1003 - 2%
1004 - 2%
1005 - 2%
1006 - 2%
1007 - 2%
1008 - 2%
1009 - 2%
1010 - 2%
1011 - 2%
1012 - 2%
1013 - 2%
1014 - 3%
1015 - 3%
1016 - 2%
1017 - 2%
1018 - 2%
1019 - 2%
1020 - 2%
1021 - 2%
1022 - 2%
1023 - 2%
1024 - 2%
1025 - 3%
1026 - 3%
1027 - 2%
1028 - 3%
1029 - 2%
1030 - 2%
1031 - 2%
1032 - 2%
1033 - 2%
103

1957 - 2%
1958 - 3%
1959 - 2%
1960 - 2%
1961 - 2%
1962 - 3%
1963 - 3%
1964 - 3%
1965 - 3%
1966 - 2%
1967 - 2%
1968 - 2%
1969 - 2%
1970 - 3%
1971 - 2%
1972 - 2%
1973 - 3%
1974 - 3%
1975 - 2%
1976 - 2%
1977 - 3%
1978 - 2%
1979 - 3%
1980 - 3%
1981 - 2%
1982 - 2%
1983 - 2%
1984 - 2%
1985 - 4%
1986 - 2%
1987 - 2%
1988 - 2%
1989 - 2%
1990 - 2%
1991 - 2%
1992 - 3%
1993 - 2%
1994 - 2%
1995 - 2%
1996 - 2%
1997 - 2%
1998 - 2%
1999 - 3%
2000 - 2%
2001 - 2%
2002 - 3%
2003 - 2%
2004 - 3%
2005 - 2%
2006 - 2%
2007 - 2%
2008 - 3%
2009 - 2%
2010 - 2%
2011 - 3%
2012 - 2%
2013 - 2%
2014 - 3%
2015 - 2%
2016 - 2%
2017 - 2%
2018 - 3%
2019 - 3%
2020 - 2%
2021 - 2%
2022 - 2%
2023 - 2%
2024 - 2%
2025 - 2%
2026 - 2%
2027 - 2%
2028 - 2%
2029 - 3%
2030 - 2%
2031 - 2%
2032 - 2%
2033 - 2%
2034 - 3%
2035 - 2%
2036 - 3%
2037 - 2%
2038 - 3%
2039 - 3%
2040 - 2%
2041 - 3%
2042 - 2%
2043 - 3%
2044 - 2%
2045 - 13%
2046 - 3%
2047 - 2%
2048 - 2%
2049 - 2%
2050 - 2%
2051 - 2%
2052 - 2%
2053 - 2%
2054 - 3%
2055 - 2%
2056 - 2%

2825 - 2%
2826 - 2%
2827 - 2%
2828 - 2%
2829 - 10%
2830 - 2%
2831 - 3%
2832 - 3%
2833 - 5%
2834 - 2%
2835 - 2%
2836 - 3%
2837 - 2%
2838 - 2%
2839 - 2%
2840 - 2%
2841 - 2%
2842 - 2%
2843 - 3%
2844 - 2%
2845 - 2%
2846 - 4%
2847 - 10%
2848 - 3%
2849 - 2%
2850 - 2%
2851 - 2%
2852 - 2%
2853 - 2%
2854 - 2%
2855 - 2%
2856 - 3%
2857 - 2%
2858 - 2%
2859 - 2%
2860 - 2%
2861 - 2%
2862 - 2%
2863 - 2%
2864 - 2%
2865 - 2%
2866 - 3%
2867 - 2%
2868 - 2%
2869 - 2%
2870 - 2%
2871 - 3%
2872 - 5%
2873 - 3%
2874 - 3%
2875 - 2%
2876 - 3%
2877 - 11%
2878 - 2%
2879 - 3%
2880 - 2%
2881 - 2%
2882 - 3%
2883 - 3%
2884 - 2%
2885 - 2%
2886 - 2%
2887 - 2%
2888 - 2%
2889 - 4%
2890 - 2%
2891 - 7%
2892 - 9%
2893 - 2%
2894 - 2%
2895 - 3%
2896 - 2%
2897 - 2%
2898 - 2%
2899 - 3%
2900 - 11%
2901 - 7%
2902 - 3%
2903 - 3%
2904 - 5%
2905 - 2%
2906 - 4%
2907 - 2%
2908 - 3%
2909 - 2%
2910 - 3%
2911 - 2%
2912 - 2%
2913 - 3%
2914 - 3%
2915 - 2%
2916 - 2%
2917 - 2%
2918 - 2%
2919 - 2%
2920 - 2%
2921 - 2%
2922 - 3%
2923 - 2%
2924 -

3727 - 7%
3728 - 11%
3729 - 7%
3730 - 3%
3731 - 5%
3732 - 2%
3733 - 2%
3734 - 6%
3735 - 2%
3736 - 4%
3737 - 2%
3738 - 4%
3739 - 6%
3740 - 7%
3741 - 12%
3742 - 11%
3743 - 9%
3744 - 4%
3745 - 2%
3746 - 3%
3747 - 2%
3748 - 2%
3749 - 2%
3750 - 2%
3751 - 6%
3752 - 7%
3753 - 6%
3754 - 13%
3755 - 4%
3756 - 11%
3757 - 11%
3758 - 2%
3759 - 12%
3760 - 3%
3761 - 2%
3762 - 13%
3763 - 2%
3764 - 2%
3765 - 13%
3766 - 10%
3767 - 2%
3768 - 11%
3769 - 11%
3770 - 10%
3771 - 11%
3772 - 2%
3773 - 3%
3774 - 13%
3775 - 13%
3776 - 14%
3777 - 12%
3778 - 4%
3779 - 4%
3780 - 11%
3781 - 2%
3782 - 2%
3783 - 2%
3784 - 11%
3785 - 2%
3786 - 11%
3787 - 9%
3788 - 2%
3789 - 2%
3790 - 3%
3791 - 13%
3792 - 11%
3793 - 6%
3794 - 2%
3795 - 2%
3796 - 9%
3797 - 13%
3798 - 2%
3799 - 13%
3800 - 2%
3801 - 2%
3802 - 4%
3803 - 4%
3804 - 2%
3805 - 2%
3806 - 13%
3807 - 11%
3808 - 11%
3809 - 9%
3810 - 2%
3811 - 12%
3812 - 3%
3813 - 2%
3814 - 2%
3815 - 11%
3816 - 2%
3817 - 3%
3818 - 2%
3819 - 2%
3820 - 7%
3821 - 7%
3822 - 10%
3823 - 2%

4571 - 2%
4572 - 2%
4573 - 2%
4574 - 2%
4575 - 2%
4576 - 2%
4577 - 2%
4578 - 2%
4579 - 2%
4580 - 2%
4581 - 5%
4582 - 9%
4583 - 14%
4584 - 11%
4585 - 10%
4586 - 11%
4587 - 11%
4588 - 2%
4589 - 14%
4590 - 2%
4591 - 2%
4592 - 4%
4593 - 13%
4594 - 12%
4595 - 2%
4596 - 7%
4597 - 3%
4598 - 5%
4599 - 11%
4600 - 12%
4601 - 13%
4602 - 11%
4603 - 5%
4604 - 14%
4605 - 5%
4606 - 10%
4607 - 7%
4608 - 2%
4609 - 2%
4610 - 12%
4611 - 5%
4612 - 12%
4613 - 10%
4614 - 2%
4615 - 7%
4616 - 3%
4617 - 12%
4618 - 4%
4619 - 9%
4620 - 3%
4621 - 3%
4622 - 2%
4623 - 2%
4624 - 10%
4625 - 12%
4626 - 10%
4627 - 12%
4628 - 4%
4629 - 10%
4630 - 13%
4631 - 2%
4632 - 11%
4633 - 2%
4634 - 4%
4635 - 8%
4636 - 11%
4637 - 3%
4638 - 11%
4639 - 3%
4640 - 2%
4641 - 12%
4642 - 7%
4643 - 2%
4644 - 5%
4645 - 7%
4646 - 4%
4647 - 12%
4648 - 11%
4649 - 11%
4650 - 11%
4651 - 3%
4652 - 3%
4653 - 2%
4654 - 2%
4655 - 11%
4656 - 2%
4657 - 9%
4658 - 4%
4659 - 2%
4660 - 2%
4661 - 4%
4662 - 2%
4663 - 3%
4664 - 3%
4665 - 11%
4666 - 2%
4667 -

5432 - 9%
5433 - 2%
5434 - 4%
5435 - 4%
5436 - 11%
5437 - 3%
5438 - 12%
5439 - 2%
5440 - 4%
5441 - 7%
5442 - 11%
5443 - 6%
5444 - 4%
5445 - 7%
5446 - 2%
5447 - 4%
5448 - 2%
5449 - 4%
5450 - 2%
5451 - 2%
5452 - 4%
5453 - 11%
5454 - 11%
5455 - 13%
5456 - 2%
5457 - 3%
5458 - 3%
5459 - 2%
5460 - 3%
5461 - 5%
5462 - 4%
5463 - 7%
5464 - 11%
5465 - 4%
5466 - 14%
5467 - 2%
5468 - 3%
5469 - 11%
5470 - 2%
5471 - 11%
5472 - 2%
5473 - 3%
5474 - 7%
5475 - 12%
5476 - 12%
5477 - 11%
5478 - 11%
5479 - 2%
5480 - 4%
5481 - 10%
5482 - 2%
5483 - 2%
5484 - 11%
5485 - 12%
5486 - 2%
5487 - 12%
5488 - 10%
5489 - 2%
5490 - 12%
5491 - 5%
5492 - 10%
5493 - 2%
5494 - 9%
5495 - 12%
5496 - 2%
5497 - 14%
5498 - 2%
5499 - 11%
5500 - 10%
5501 - 3%
5502 - 3%
5503 - 11%
5504 - 4%
5505 - 5%
5506 - 2%
5507 - 14%
5508 - 11%
5509 - 9%
5510 - 12%
5511 - 12%
5512 - 2%
5513 - 12%
5514 - 4%
5515 - 10%
5516 - 9%
5517 - 4%
5518 - 3%
5519 - 2%
5520 - 12%
5521 - 4%
5522 - 10%
5523 - 2%
5524 - 2%
5525 - 4%
5526 - 2%
5527 - 11%
5528 

6263 - 2%
6264 - 6%
6265 - 2%
6266 - 6%
6267 - 3%
6268 - 2%
6269 - 14%
6270 - 3%
6271 - 11%
6272 - 3%
6273 - 5%
6274 - 3%
6275 - 2%
6276 - 7%
6277 - 4%
6278 - 2%
6279 - 3%
6280 - 10%
6281 - 2%
6282 - 4%
6283 - 3%
6284 - 2%
6285 - 2%
6286 - 4%
6287 - 2%
6288 - 2%
6289 - 11%
6290 - 12%
6291 - 7%
6292 - 12%
6293 - 2%
6294 - 3%
6295 - 4%
6296 - 5%
6297 - 2%
6298 - 2%
6299 - 5%
6300 - 3%
6301 - 9%
6302 - 2%
6303 - 4%
6304 - 2%
6305 - 3%
6306 - 3%
6307 - 2%
6308 - 12%
6309 - 3%
6310 - 2%
6311 - 2%
6312 - 12%
6313 - 2%
6314 - 3%
6315 - 2%
6316 - 12%
6317 - 12%
6318 - 3%
6319 - 2%
6320 - 2%
6321 - 2%
6322 - 3%
6323 - 2%
6324 - 2%
6325 - 4%
6326 - 3%
6327 - 13%
6328 - 3%
6329 - 2%
6330 - 4%
6331 - 2%
6332 - 7%
6333 - 2%
6334 - 13%
6335 - 7%
6336 - 12%
6337 - 2%
6338 - 2%
6339 - 2%
6340 - 11%
6341 - 2%
6342 - 4%
6343 - 10%
6344 - 11%
6345 - 6%
6346 - 11%
6347 - 5%
6348 - 3%
6349 - 2%
6350 - 4%
6351 - 10%
6352 - 5%
6353 - 14%
6354 - 2%
6355 - 10%
6356 - 12%
6357 - 7%
6358 - 2%
6359 - 2%
6360 - 2%

7082 - 3%
7083 - 2%
7084 - 3%
7085 - 4%
7086 - 12%
7087 - 2%
7088 - 2%
7089 - 7%
7090 - 2%
7091 - 12%
7092 - 10%
7093 - 4%
7094 - 2%
7095 - 2%
7096 - 5%
7097 - 2%
7098 - 2%
7099 - 12%
7100 - 2%
7101 - 3%
7102 - 2%
7103 - 12%
7104 - 12%
7105 - 4%
7106 - 13%
7107 - 14%
7108 - 2%
7109 - 2%
7110 - 2%
7111 - 12%
7112 - 2%
7113 - 3%
7114 - 12%
7115 - 2%
7116 - 3%
7117 - 2%
7118 - 2%
7119 - 12%
7120 - 4%
7121 - 4%
7122 - 2%
7123 - 2%
7124 - 4%
7125 - 2%
7126 - 2%
7127 - 2%
7128 - 2%
7129 - 2%
7130 - 2%
7131 - 2%
7132 - 3%
7133 - 2%
7134 - 7%
7135 - 11%
7136 - 3%
7137 - 2%
7138 - 2%
7139 - 2%
7140 - 2%
7141 - 2%
7142 - 12%
7143 - 4%
7144 - 12%
7145 - 4%
7146 - 12%
7147 - 2%
7148 - 2%
7149 - 12%
7150 - 3%
7151 - 2%
7152 - 4%
7153 - 7%
7154 - 4%
7155 - 7%
7156 - 2%
7157 - 2%
7158 - 3%
7159 - 10%
7160 - 2%
7161 - 7%
7162 - 2%
7163 - 11%
7164 - 2%
7165 - 2%
7166 - 2%
7167 - 2%
7168 - 7%
7169 - 12%
7170 - 3%
7171 - 2%
7172 - 3%
7173 - 10%
7174 - 2%
7175 - 14%
7176 - 12%
7177 - 2%
7178 - 3%
7179 - 2

8017 - 12%
8018 - 5%
8019 - 5%
8020 - 2%
8021 - 5%
8022 - 2%
8023 - 12%
8024 - 5%
8025 - 13%
8026 - 4%
8027 - 2%
8028 - 7%
8029 - 5%
8030 - 11%
8031 - 2%
8032 - 5%
8033 - 2%
8034 - 2%
8035 - 2%
8036 - 2%
8037 - 2%
8038 - 2%
8039 - 14%
8040 - 10%
8041 - 4%
8042 - 10%
8043 - 2%
8044 - 11%
8045 - 7%
8046 - 3%
8047 - 2%
8048 - 5%
8049 - 4%
8050 - 6%
8051 - 2%
8052 - 2%
8053 - 5%
8054 - 12%
8055 - 12%
8056 - 11%
8057 - 2%
8058 - 4%
8060 - 11%
8061 - 12%
8062 - 12%
8063 - 2%
8064 - 12%
8065 - 5%
8066 - 4%
8067 - 10%
8068 - 11%
8069 - 2%
8070 - 4%
8071 - 7%
8072 - 3%
8073 - 3%
8074 - 7%
8075 - 2%
8076 - 2%
8077 - 2%
8078 - 5%
8079 - 4%
8080 - 10%
8081 - 3%
8082 - 12%
8083 - 3%
8084 - 2%
8085 - 7%
8086 - 12%
8087 - 4%
8088 - 2%
8089 - 3%
8090 - 4%
8091 - 7%
8092 - 5%
8093 - 2%
8094 - 2%
8095 - 2%
8096 - 6%
8097 - 12%
8098 - 11%
8099 - 2%
8100 - 5%
8101 - 2%
8102 - 2%
8103 - 4%
8104 - 5%
8105 - 5%
8106 - 11%
8107 - 11%
8108 - 7%
8109 - 2%
8110 - 3%
8111 - 11%
8112 - 4%
8113 - 4%
8114 - 9%
8115 

9112 - 7%
9114 - 2%
9115 - 2%
9116 - 1%
9117 - 0%
9118 - 5%
9119 - 4%
9120 - 0%
9121 - 1%
9122 - 2%
9123 - 14%
9124 - 2%
9125 - 4%
9126 - 2%
9128 - 2%
9130 - 10%
9131 - 2%
9132 - 5%
9133 - 3%
9134 - 2%
9135 - 11%
9136 - 3%
9137 - 0%
9138 - 2%
9139 - 4%
9140 - 2%
9141 - 3%
9142 - 0%
9143 - 2%
9144 - 4%
9145 - 2%
9146 - 2%
9148 - 2%
9149 - 2%
9150 - 5%
9151 - 2%
9152 - 2%
9153 - 2%
9154 - 3%
9156 - 3%
9157 - 12%
9158 - 1%
9159 - 2%
9160 - 2%
9162 - 3%
9163 - 0%
9165 - 3%
9166 - 4%
9167 - 5%
9168 - 2%
9169 - 5%
9170 - 0%
9171 - 3%
9172 - 2%
9174 - 2%
9175 - 0%
9176 - 5%
9177 - 2%
9179 - 5%
9180 - 2%
9181 - 2%
9182 - 5%
9183 - 0%
9184 - 1%
9185 - 2%
9186 - 5%
9187 - 2%
9188 - 2%
9189 - 2%
9190 - 3%
9191 - 2%
9192 - 5%
9193 - 1%
9194 - 3%
9195 - 10%
9196 - 2%
9197 - 2%
9198 - 3%
9199 - 2%
9200 - 5%
9201 - 1%
9202 - 1%
9203 - 11%
9204 - 2%
9205 - 14%
9206 - 14%
9207 - 0%
9209 - 4%
9210 - 3%
9212 - 2%
9213 - 14%
9214 - 2%
9215 - 2%
9216 - 1%
9217 - 10%
9218 - 13%
9219 - 11%
9220 - 4%
9221 - 4

10221 - 4%
10222 - 4%
10224 - 1%
10225 - 0%
10226 - 0%
10227 - 4%
10229 - 1%
10230 - 2%
10232 - 12%
10233 - 2%
10234 - 5%
10235 - 4%
10236 - 3%
10237 - 0%
10238 - 1%
10239 - 0%
10241 - 0%
10242 - 3%
10243 - 1%
10244 - 1%
10246 - 1%
10247 - 3%
10248 - 0%
10250 - 1%
10254 - 2%
10255 - 0%
10256 - 0%
10257 - 2%
10258 - 9%
10259 - 0%
10260 - 8%
10261 - 2%
10262 - 10%
10263 - 2%
10264 - 0%
10265 - 1%
10266 - 2%
10268 - 8%
10269 - 10%
10270 - 9%
10272 - 3%
10273 - 0%
10274 - 1%
10278 - 3%
10279 - 9%
10280 - 0%
10281 - 1%
10282 - 2%
10283 - 0%
10284 - 5%
10285 - 0%
10286 - 1%
10287 - 2%
10289 - 0%
10290 - 9%
10291 - 3%
10292 - 0%
10293 - 3%
10294 - 0%
10295 - 1%
10296 - 3%
10297 - 3%
10300 - 1%
10302 - 0%
10303 - 0%
10304 - 8%
10305 - 0%
10307 - 11%
10308 - 4%
10309 - 4%
10310 - 0%
10311 - 1%
10312 - 0%
10313 - 0%
10315 - 1%
10316 - 3%
10319 - 3%
10321 - 2%
10322 - 0%
10323 - 0%
10324 - 3%
10325 - 2%
10326 - 4%
10327 - 0%
10328 - 0%
10329 - 1%
10331 - 2%
10332 - 1%
10333 - 10%
10334 - 0%
10335

11365 - 0%
11367 - 0%
11368 - 2%
11369 - 1%
11370 - 5%
11371 - 0%
11373 - 1%
11375 - 1%
11376 - 1%
11378 - 2%
11382 - 5%
11383 - 0%
11384 - 3%
11385 - 1%
11386 - 12%
11388 - 0%
11390 - 1%
11393 - 9%
11394 - 9%
11395 - 1%
11396 - 1%
11397 - 10%
11398 - 3%
11399 - 4%
11400 - 3%
11401 - 1%
11402 - 1%
11403 - 1%
11404 - 1%
11407 - 9%
11408 - 4%
11409 - 10%
11413 - 1%
11414 - 3%
11416 - 4%
11417 - 0%
11418 - 0%
11421 - 1%
11422 - 0%
11423 - 0%
11425 - 7%
11426 - 3%
11429 - 0%
11430 - 0%
11434 - 1%
11435 - 4%
11436 - 1%
11437 - 8%
11438 - 2%
11439 - 1%
11440 - 0%
11442 - 1%
11446 - 4%
11447 - 0%
11448 - 4%
11450 - 1%
11451 - 4%
11452 - 0%
11453 - 0%
11454 - 5%
11455 - 2%
11456 - 0%
11458 - 0%
11462 - 1%
11463 - 2%
11465 - 4%
11466 - 2%
11467 - 2%
11468 - 0%
11470 - 0%
11471 - 3%
11472 - 0%
11473 - 1%
11476 - 4%
11477 - 4%
11480 - 4%
11481 - 8%
11484 - 0%
11485 - 8%
11486 - 1%
11487 - 1%
11489 - 0%
11491 - 0%
11492 - 0%
11493 - 4%
11494 - 9%
11495 - 7%
11496 - 3%
11498 - 5%
11500 - 9%
11502 -

12470 - 9%
12471 - 4%
12472 - 0%
12473 - 0%
12475 - 2%
12476 - 7%
12477 - 0%
12481 - 0%
12482 - 1%
12483 - 0%
12484 - 4%
12485 - 1%
12486 - 1%
12487 - 3%
12488 - 1%
12490 - 2%
12493 - 1%
12494 - 3%
12496 - 1%
12497 - 1%
12498 - 1%
12500 - 4%
12504 - 4%
12505 - 4%
12506 - 1%
12507 - 3%
12508 - 1%
12511 - 3%
12513 - 2%
12514 - 0%
12515 - 12%
12517 - 3%
12518 - 12%
12519 - 0%
12520 - 8%
12522 - 9%
12523 - 2%
12524 - 0%
12525 - 1%
12526 - 1%
12527 - 1%
12528 - 1%
12530 - 3%
12533 - 0%
12534 - 4%
12535 - 1%
12537 - 3%
12539 - 3%
12540 - 4%
12541 - 1%
12542 - 1%
12544 - 12%
12545 - 1%
12546 - 3%
12548 - 3%
12549 - 12%
12550 - 1%
12553 - 8%
12554 - 4%
12558 - 9%
12560 - 9%
12561 - 12%
12562 - 0%
12564 - 8%
12565 - 1%
12566 - 1%
12567 - 4%
12571 - 1%
12572 - 0%
12574 - 0%
12575 - 0%
12577 - 1%
12578 - 0%
12579 - 1%
12580 - 1%
12581 - 1%
12583 - 4%
12585 - 1%
12586 - 1%
12587 - 0%
12589 - 4%
12590 - 1%
12593 - 1%
12597 - 0%
12599 - 1%
12602 - 1%
12603 - 10%
12605 - 4%
12606 - 5%
12607 - 1%
1260

13589 - 2%
13590 - 7%
13591 - 1%
13593 - 4%
13594 - 1%
13598 - 10%
13599 - 0%
13600 - 1%
13601 - 1%
13603 - 2%
13604 - 0%
13607 - 3%
13610 - 4%
13611 - 1%
13612 - 1%
13613 - 3%
13616 - 1%
13618 - 3%
13619 - 9%
13620 - 9%
13622 - 7%
13623 - 0%
13624 - 3%
13626 - 9%
13630 - 2%
13632 - 2%
13633 - 0%
13634 - 1%
13636 - 2%
13637 - 0%
13638 - 2%
13639 - 1%
13640 - 0%
13642 - 1%
13643 - 2%
13644 - 1%
13645 - 10%
13646 - 1%
13647 - 1%
13648 - 4%
13650 - 4%
13651 - 0%
13652 - 1%
13653 - 1%
13654 - 3%
13655 - 4%
13656 - 10%
13657 - 11%
13658 - 3%
13660 - 8%
13663 - 1%
13664 - 2%
13666 - 0%
13667 - 0%
13668 - 0%
13669 - 9%
13671 - 2%
13672 - 9%
13674 - 0%
13675 - 7%
13676 - 1%
13677 - 1%
13678 - 4%
13679 - 11%
13680 - 0%
13681 - 1%
13683 - 1%
13684 - 1%
13685 - 1%
13686 - 1%
13687 - 4%
13688 - 0%
13689 - 4%
13690 - 1%
13692 - 0%
13693 - 1%
13694 - 12%
13695 - 1%
13696 - 4%
13697 - 0%
13698 - 4%
13699 - 0%
13700 - 2%
13701 - 1%
13702 - 2%
13703 - 1%
13704 - 7%
13706 - 3%
13707 - 0%
13708 - 0%
1370

14725 - 0%
14727 - 12%
14729 - 4%
14730 - 1%
14731 - 3%
14732 - 8%
14733 - 0%
14734 - 0%
14735 - 0%
14736 - 2%
14737 - 13%
14738 - 0%
14739 - 10%
14741 - 2%
14743 - 0%
14745 - 0%
14746 - 0%
14747 - 1%
14749 - 1%
14750 - 13%
14752 - 0%
14757 - 0%
14758 - 1%
14759 - 0%
14761 - 1%
14762 - 1%
14763 - 8%
14768 - 1%
14769 - 2%
14770 - 0%
14771 - 13%
14772 - 12%
14774 - 9%
14775 - 0%
14776 - 12%
14777 - 3%
14779 - 3%
14782 - 0%
14784 - 0%
14785 - 1%
14787 - 1%
14788 - 12%
14789 - 3%
14790 - 4%
14792 - 7%
14794 - 1%
14797 - 0%
14799 - 2%
14802 - 3%
14803 - 1%
14804 - 0%
14805 - 0%
14807 - 0%
14808 - 1%
14810 - 4%
14811 - 0%
14813 - 3%
14814 - 7%
14815 - 10%
14817 - 1%
14818 - 4%
14819 - 1%
14820 - 0%
14821 - 3%
14822 - 3%
14823 - 0%
14824 - 8%
14825 - 7%
14826 - 1%
14827 - 1%
14829 - 13%
14830 - 1%
14831 - 7%
14832 - 1%
14833 - 0%
14834 - 1%
14835 - 8%
14837 - 1%
14838 - 1%
14839 - 3%
14840 - 1%
14841 - 1%
14842 - 2%
14843 - 0%
14844 - 3%
14846 - 8%
14847 - 1%
14849 - 9%
14851 - 0%
14852 - 1%


16082 - 1%
16083 - 0%
16084 - 7%
16085 - 0%
16086 - 1%
16087 - 0%
16088 - 0%
16089 - 1%
16090 - 7%
16092 - 1%
16093 - 12%
16094 - 0%
16095 - 3%
16097 - 1%
16099 - 2%
16100 - 2%
16101 - 1%
16103 - 1%
16104 - 1%
16106 - 1%
16107 - 2%
16108 - 2%
16109 - 1%
16110 - 1%
16111 - 12%
16112 - 3%
16113 - 1%
16115 - 0%
16116 - 0%
16118 - 2%
16119 - 8%
16120 - 0%
16124 - 8%
16126 - 1%
16127 - 0%
16128 - 4%
16130 - 10%
16131 - 1%
16132 - 1%
16133 - 0%
16134 - 1%
16135 - 3%
16136 - 1%
16137 - 1%
16138 - 1%
16140 - 2%
16141 - 3%
16142 - 1%
16143 - 3%
16144 - 1%
16146 - 2%
16147 - 1%
16149 - 8%
16150 - 0%
16151 - 2%
16152 - 0%
16153 - 2%
16154 - 1%
16155 - 0%
16156 - 4%
16157 - 1%
16159 - 1%
16160 - 0%
16161 - 7%
16164 - 0%
16166 - 13%
16167 - 0%
16168 - 7%
16169 - 1%
16170 - 7%
16171 - 1%
16173 - 0%
16175 - 3%
16176 - 1%
16179 - 14%
16180 - 4%
16181 - 0%
16183 - 1%
16186 - 10%
16187 - 9%
16188 - 1%
16189 - 1%
16190 - 12%
16191 - 3%
16193 - 5%
16194 - 0%
16195 - 1%
16196 - 1%
16197 - 1%
16198 - 1%
161

17196 - 2%
17198 - 8%
17199 - 1%
17201 - 2%
17202 - 1%
17203 - 2%
17204 - 1%
17205 - 0%
17208 - 3%
17209 - 0%
17214 - 0%
17216 - 3%
17217 - 0%
17218 - 1%
17219 - 7%
17220 - 7%
17221 - 1%
17222 - 1%
17223 - 7%
17226 - 1%
17227 - 1%
17228 - 3%
17229 - 8%
17230 - 9%
17231 - 0%
17232 - 2%
17233 - 1%
17234 - 1%
17235 - 1%
17236 - 0%
17237 - 1%
17238 - 7%
17239 - 0%
17240 - 1%
17241 - 2%
17242 - 1%
17243 - 1%
17245 - 1%
17246 - 3%
17248 - 4%
17249 - 0%
17251 - 2%
17252 - 13%
17258 - 1%
17259 - 3%
17262 - 2%
17264 - 1%
17265 - 0%
17266 - 2%
17267 - 1%
17269 - 1%
17272 - 1%
17275 - 3%
17277 - 0%
17278 - 2%
17279 - 2%
17280 - 0%
17281 - 8%
17282 - 0%
17283 - 1%
17284 - 1%
17285 - 2%
17286 - 0%
17287 - 0%
17290 - 4%
17291 - 2%
17292 - 1%
17293 - 0%
17294 - 8%
17295 - 8%
17299 - 0%
17300 - 2%
17302 - 3%
17303 - 9%
17304 - 0%
17305 - 1%
17306 - 0%
17308 - 3%
17309 - 4%
17310 - 1%
17312 - 8%
17313 - 0%
17315 - 3%
17316 - 9%
17317 - 12%
17319 - 2%
17320 - 1%
17321 - 0%
17322 - 8%
17323 - 0%
17324 - 

18237 - 12%
18238 - 0%
18239 - 0%
18240 - 2%
18241 - 1%
18242 - 10%
18243 - 4%
18246 - 1%
18247 - 0%
18248 - 0%
18249 - 0%
18250 - 2%
18251 - 3%
18252 - 12%
18254 - 3%
18255 - 0%
18257 - 3%
18258 - 2%
18259 - 1%
18260 - 2%
18262 - 4%
18263 - 9%
18265 - 4%
18266 - 0%
18270 - 3%
18271 - 0%
18273 - 1%
18276 - 13%
18277 - 14%
18278 - 3%
18279 - 0%
18280 - 3%
18283 - 12%
18285 - 4%
18287 - 1%
18290 - 1%
18291 - 0%
18292 - 8%
18294 - 8%
18295 - 4%
18296 - 2%
18298 - 0%
18299 - 1%
18301 - 0%
18303 - 1%
18304 - 0%
18305 - 4%
18306 - 4%
18307 - 4%
18309 - 8%
18310 - 0%
18311 - 4%
18312 - 4%
18314 - 1%
18316 - 9%
18317 - 0%
18318 - 0%
18319 - 5%
18320 - 0%
18321 - 0%
18322 - 1%
18323 - 0%
18325 - 9%
18326 - 1%
18327 - 0%
18328 - 1%
18329 - 3%
18330 - 1%
18331 - 8%
18332 - 0%
18333 - 1%
18334 - 2%
18335 - 3%
18336 - 2%
18337 - 13%
18338 - 1%
18339 - 4%
18340 - 1%
18342 - 0%
18343 - 1%
18344 - 4%
18346 - 0%
18350 - 4%
18351 - 0%
18353 - 4%
18355 - 2%
18356 - 2%
18357 - 1%
18358 - 0%
18359 - 8%
183

19422 - 4%
19423 - 0%
19424 - 1%
19425 - 0%
19426 - 7%
19427 - 3%
19429 - 0%
19431 - 1%
19432 - 1%
19433 - 2%
19434 - 3%
19435 - 1%
19436 - 2%
19439 - 12%
19440 - 0%
19441 - 0%
19442 - 0%
19443 - 8%
19444 - 3%
19445 - 9%
19447 - 2%
19449 - 1%
19450 - 8%
19451 - 0%
19452 - 8%
19453 - 1%
19455 - 0%
19459 - 0%
19460 - 1%
19461 - 1%
19463 - 0%
19464 - 7%
19465 - 0%
19466 - 10%
19467 - 8%
19468 - 0%
19469 - 1%
19470 - 0%
19472 - 1%
19473 - 7%
19474 - 1%
19475 - 0%
19476 - 0%
19477 - 1%
19478 - 4%
19479 - 9%
19480 - 2%
19481 - 4%
19482 - 8%
19483 - 1%
19484 - 8%
19485 - 1%
19487 - 8%
19488 - 1%
19489 - 7%
19490 - 2%
19491 - 0%
19492 - 4%
19494 - 11%
19495 - 0%
19496 - 8%
19497 - 3%
19498 - 12%
19499 - 1%
19500 - 1%
19501 - 1%
19502 - 2%
19504 - 7%
19506 - 0%
19507 - 7%
19508 - 1%
19510 - 0%
19512 - 2%
19514 - 2%
19515 - 0%
19517 - 1%
19518 - 1%
19520 - 0%
19521 - 0%
19522 - 1%
19523 - 9%
19524 - 12%
19525 - 1%
19526 - 12%
19527 - 10%
19530 - 0%
19531 - 1%
19532 - 1%
19535 - 8%
19536 - 1%
195

20468 - 0%
20470 - 0%
20471 - 1%
20472 - 1%
20474 - 9%
20476 - 1%
20478 - 1%
20479 - 7%
20480 - 1%
20481 - 1%
20482 - 4%
20484 - 5%
20486 - 1%
20487 - 2%
20489 - 3%
20490 - 1%
20492 - 0%
20493 - 0%
20494 - 3%
20496 - 8%
20498 - 13%
20499 - 0%
20500 - 1%
20501 - 0%
20502 - 1%
20504 - 0%
20506 - 2%
20507 - 7%
20508 - 2%
20509 - 1%
20510 - 0%
20511 - 1%
20512 - 3%
20515 - 0%
20517 - 0%
20520 - 3%
20521 - 8%
20522 - 1%
20523 - 1%
20525 - 0%
20526 - 0%
20527 - 2%
20528 - 8%
20530 - 0%
20531 - 8%
20532 - 0%
20533 - 8%
20534 - 0%
20536 - 1%
20538 - 1%
20540 - 10%
20542 - 1%
20543 - 0%
20544 - 3%
20545 - 1%
20546 - 12%
20547 - 1%
20549 - 3%
20552 - 1%
20553 - 1%
20554 - 1%
20555 - 3%
20556 - 1%
20557 - 0%
20558 - 3%
20560 - 8%
20562 - 8%
20563 - 4%
20564 - 7%
20565 - 4%
20566 - 1%
20568 - 1%
20569 - 4%
20570 - 1%
20572 - 2%
20573 - 5%
20574 - 8%
20575 - 1%
20576 - 12%
20577 - 1%
20578 - 3%
20579 - 0%
20581 - 3%
20582 - 3%
20583 - 0%
20584 - 1%
20585 - 1%
20586 - 1%
20587 - 7%
20588 - 1%
20589 

21638 - 0%
21639 - 3%
21641 - 8%
21642 - 1%
21643 - 4%
21644 - 8%
21646 - 8%
21648 - 1%
21649 - 1%
21651 - 8%
21652 - 7%
21653 - 3%
21654 - 3%
21655 - 3%
21656 - 1%
21657 - 0%
21658 - 1%
21659 - 0%
21661 - 0%
21662 - 2%
21663 - 1%
21664 - 12%
21665 - 0%
21666 - 8%
21667 - 8%
21669 - 1%
21670 - 1%
21672 - 0%
21673 - 3%
21674 - 1%
21675 - 1%
21676 - 9%
21677 - 1%
21678 - 0%
21679 - 7%
21680 - 7%
21681 - 4%
21682 - 1%
21683 - 1%
21684 - 2%
21685 - 3%
21686 - 1%
21690 - 1%
21691 - 0%
21692 - 1%
21693 - 13%
21694 - 1%
21695 - 0%
21696 - 1%
21697 - 1%
21698 - 5%
21699 - 0%
21700 - 0%
21701 - 8%
21702 - 11%
21703 - 1%
21704 - 0%
21705 - 1%
21706 - 9%
21707 - 1%
21708 - 1%
21710 - 1%
21711 - 3%
21712 - 1%
21714 - 8%
21715 - 0%
21716 - 1%
21717 - 1%
21720 - 2%
21721 - 2%
21722 - 0%
21723 - 1%
21724 - 0%
21725 - 2%
21726 - 1%
21727 - 1%
21728 - 1%
21729 - 1%
21730 - 2%
21731 - 8%
21732 - 3%
21733 - 0%
21736 - 8%
21737 - 3%
21738 - 3%
21741 - 1%
21742 - 1%
21743 - 0%
21745 - 0%
21746 - 1%
21747 -

22763 - 13%
22765 - 3%
22769 - 12%
22776 - 1%
22777 - 0%
22778 - 3%
22779 - 1%
22783 - 0%
22784 - 4%
22785 - 1%
22787 - 12%
22789 - 8%
22790 - 8%
22791 - 0%
22792 - 0%
22795 - 8%
22796 - 9%
22798 - 3%
22799 - 1%
22800 - 7%
22801 - 1%
22802 - 1%
22803 - 1%
22806 - 7%
22807 - 1%
22809 - 1%
22810 - 0%
22811 - 1%
22812 - 8%
22813 - 7%
22814 - 0%
22815 - 1%
22816 - 12%
22817 - 2%
22818 - 13%
22820 - 1%
22821 - 1%
22822 - 7%
22823 - 1%
22824 - 1%
22825 - 0%
22827 - 2%
22828 - 0%
22829 - 3%
22832 - 1%
22834 - 1%
22836 - 1%
22838 - 0%
22839 - 8%
22840 - 1%
22841 - 3%
22842 - 1%
22843 - 3%
22849 - 1%
22850 - 12%
22851 - 2%
22852 - 0%
22854 - 0%
22855 - 8%
22858 - 0%
22859 - 0%
22860 - 0%
22861 - 0%
22862 - 1%
22863 - 14%
22864 - 0%
22865 - 1%
22866 - 2%
22867 - 0%
22868 - 0%
22869 - 0%
22870 - 3%
22871 - 3%
22872 - 3%
22873 - 7%
22874 - 3%
22875 - 1%
22876 - 0%
22880 - 1%
22881 - 3%
22882 - 8%
22883 - 0%
22884 - 1%
22885 - 8%
22887 - 1%
22889 - 1%
22890 - 1%
22892 - 0%
22893 - 0%
22894 - 0%
228

23838 - 1%
23839 - 0%
23840 - 3%
23841 - 1%
23844 - 1%
23845 - 7%
23846 - 12%
23847 - 1%
23848 - 0%
23849 - 8%
23851 - 0%
23852 - 1%
23853 - 3%
23854 - 0%
23855 - 0%
23858 - 1%
23860 - 1%
23861 - 0%
23862 - 1%
23863 - 10%
23865 - 0%
23866 - 8%
23867 - 9%
23869 - 11%
23870 - 12%
23871 - 12%
23872 - 9%
23873 - 3%
23874 - 0%
23876 - 0%
23877 - 7%
23878 - 0%
23879 - 3%
23880 - 4%
23883 - 3%
23884 - 1%
23885 - 7%
23886 - 2%
23888 - 8%
23889 - 3%
23890 - 0%
23891 - 3%
23892 - 1%
23893 - 0%
23894 - 0%
23896 - 2%
23899 - 0%
23900 - 1%
23901 - 1%
23902 - 3%
23904 - 8%
23905 - 0%
23906 - 4%
23907 - 1%
23909 - 9%
23910 - 0%
23912 - 1%
23914 - 0%
23915 - 3%
23917 - 7%
23918 - 4%
23919 - 2%
23920 - 1%
23921 - 3%
23922 - 0%
23923 - 3%
23924 - 0%
23927 - 8%
23928 - 1%
23929 - 11%
23930 - 0%
23933 - 4%
23934 - 0%
23935 - 3%
23936 - 0%
23937 - 1%
23938 - 0%
23939 - 3%
23940 - 0%
23942 - 0%
23944 - 0%
23945 - 1%
23947 - 5%
23948 - 4%
23949 - 8%
23950 - 2%
23951 - 13%
23952 - 1%
23954 - 4%
23955 - 3%
239

25079 - 0%
25080 - 3%
25081 - 1%
25083 - 1%
25084 - 0%
25086 - 0%
25087 - 12%
25090 - 8%
25093 - 2%
25094 - 4%
25095 - 2%
25096 - 13%
25097 - 8%
25099 - 4%
25100 - 0%
25101 - 1%
25102 - 0%
25105 - 14%
25106 - 0%
25109 - 0%
25110 - 10%
25111 - 1%
25112 - 7%
25113 - 1%
25115 - 3%
25116 - 0%
25117 - 0%
25118 - 1%
25120 - 0%
25121 - 3%
25123 - 7%
25124 - 2%
25125 - 0%
25126 - 8%
25127 - 1%
25131 - 0%
25132 - 1%
25134 - 0%
25136 - 1%
25137 - 8%
25140 - 1%
25141 - 0%
25142 - 13%
25143 - 8%
25144 - 1%
25145 - 2%
25146 - 1%
25148 - 2%
25149 - 7%
25150 - 8%
25151 - 1%
25153 - 3%
25154 - 7%
25155 - 1%
25156 - 0%
25157 - 1%
25158 - 2%
25160 - 1%
25161 - 8%
25163 - 3%
25164 - 10%
25165 - 8%
25166 - 8%
25167 - 2%
25168 - 0%
25170 - 0%
25171 - 3%
25172 - 2%
25173 - 0%
25174 - 1%
25178 - 0%
25179 - 0%
25181 - 1%
25182 - 4%
25183 - 0%
25187 - 3%
25190 - 1%
25191 - 0%
25192 - 0%
25194 - 0%
25195 - 0%
25196 - 3%
25198 - 1%
25200 - 12%
25202 - 0%
25204 - 7%
25205 - 0%
25206 - 13%
25208 - 1%
25209 - 3%
25

26330 - 3%
26331 - 7%
26332 - 8%
26334 - 0%
26335 - 1%
26336 - 8%
26337 - 0%
26340 - 7%
26341 - 10%
26342 - 1%
26343 - 1%
26344 - 2%
26345 - 1%
26346 - 7%
26347 - 2%
26348 - 12%
26349 - 7%
26351 - 7%
26352 - 0%
26353 - 0%
26354 - 4%
26355 - 0%
26356 - 0%
26357 - 0%
26358 - 1%
26359 - 0%
26360 - 1%
26361 - 10%
26362 - 0%
26363 - 1%
26365 - 1%
26366 - 1%
26367 - 8%
26368 - 1%
26369 - 0%
26370 - 0%
26371 - 1%
26372 - 7%
26373 - 3%
26374 - 4%
26375 - 8%
26376 - 8%
26379 - 0%
26380 - 0%
26381 - 7%
26382 - 7%
26383 - 8%
26384 - 1%
26385 - 3%
26386 - 8%
26388 - 7%
26389 - 2%
26390 - 0%
26391 - 1%
26392 - 1%
26394 - 0%
26396 - 3%
26397 - 7%
26398 - 8%
26399 - 0%
26401 - 3%
26402 - 3%
26404 - 3%
26405 - 7%
26407 - 0%
26408 - 10%
26411 - 3%
26412 - 8%
26413 - 3%
26414 - 7%
26415 - 1%
26416 - 1%
26417 - 3%
26418 - 2%
26419 - 0%
26420 - 1%
26421 - 3%
26422 - 1%
26423 - 8%
26424 - 3%
26425 - 3%
26426 - 4%
26428 - 9%
26429 - 1%
26431 - 2%
26433 - 12%
26435 - 0%
26436 - 9%
26437 - 10%
26438 - 1%
2643

27463 - 0%
27465 - 1%
27466 - 1%
27468 - 1%
27470 - 3%
27472 - 1%
27473 - 0%
27476 - 2%
27479 - 11%
27480 - 1%
27481 - 0%
27482 - 1%
27483 - 1%
27484 - 0%
27485 - 1%
27486 - 8%
27488 - 8%
27489 - 2%
27491 - 1%
27492 - 0%
27493 - 3%
27495 - 0%
27496 - 0%
27499 - 0%
27501 - 1%
27502 - 7%
27503 - 8%
27504 - 8%
27506 - 0%
27507 - 1%
27508 - 3%
27509 - 2%
27510 - 7%
27511 - 0%
27512 - 0%
27515 - 0%
27516 - 2%
27517 - 7%
27518 - 7%
27520 - 12%
27521 - 1%
27524 - 0%
27526 - 1%
27528 - 1%
27532 - 10%
27533 - 1%
27535 - 3%
27536 - 0%
27537 - 1%
27539 - 3%
27541 - 2%
27542 - 0%
27544 - 1%
27547 - 3%
27548 - 3%
27550 - 0%
27551 - 1%
27554 - 1%
27555 - 8%
27557 - 0%
27558 - 2%
27559 - 0%
27560 - 8%
27561 - 0%
27563 - 4%
27565 - 2%
27568 - 7%
27569 - 0%
27570 - 3%
27571 - 12%
27574 - 0%
27575 - 3%
27576 - 10%
27580 - 1%
27581 - 2%
27582 - 1%
27584 - 1%
27586 - 9%
27587 - 0%
27588 - 1%
27589 - 0%
27594 - 1%
27596 - 0%
27598 - 0%
27599 - 1%
27600 - 1%
27601 - 3%
27602 - 14%
27603 - 8%
27604 - 0%
2760

28693 - 7%
28694 - 7%
28695 - 1%
28696 - 4%
28697 - 1%
28698 - 7%
28699 - 14%
28700 - 8%
28702 - 3%
28703 - 7%
28704 - 7%
28705 - 4%
28706 - 0%
28709 - 10%
28710 - 12%
28711 - 11%
28712 - 11%
28713 - 2%
28714 - 1%
28715 - 11%
28716 - 14%
28717 - 11%
28719 - 8%
28720 - 7%
28721 - 3%
28722 - 3%
28723 - 1%
28724 - 7%
28725 - 1%
28726 - 3%
28728 - 9%
28729 - 7%
28731 - 0%
28732 - 1%
28733 - 1%
28734 - 13%
28735 - 1%
28736 - 0%
28737 - 0%
28738 - 1%
28739 - 8%
28741 - 2%
28743 - 4%
28744 - 8%
28745 - 0%
28747 - 1%
28748 - 0%
28750 - 1%
28751 - 0%
28752 - 13%
28754 - 1%
28755 - 8%
28756 - 1%
28757 - 8%
28758 - 14%
28759 - 2%
28760 - 7%
28761 - 7%
28762 - 0%
28763 - 7%
28764 - 8%
28765 - 0%
28766 - 14%
28767 - 7%
28768 - 1%
28769 - 1%
28770 - 14%
28772 - 10%
28773 - 8%
28775 - 5%
28776 - 0%
28777 - 1%
28779 - 0%
28780 - 1%
28781 - 1%
28783 - 0%
28786 - 0%
28787 - 0%
28788 - 3%
28789 - 1%
28790 - 0%
28791 - 0%
28792 - 4%
28793 - 7%
28794 - 3%
28796 - 0%
28797 - 0%
28798 - 1%
28799 - 1%
28801 -

29707 - 0%
29708 - 7%
29709 - 14%
29710 - 10%
29714 - 1%
29716 - 1%
29717 - 10%
29719 - 2%
29720 - 1%
29723 - 12%
29724 - 7%
29725 - 0%
29726 - 12%
29729 - 12%
29730 - 1%
29731 - 3%
29733 - 7%
29734 - 0%
29735 - 3%
29739 - 2%
29740 - 14%
29741 - 0%
29742 - 7%
29743 - 11%
29745 - 12%
29746 - 2%
29747 - 1%
29748 - 1%
29749 - 1%
29752 - 1%
29753 - 0%
29756 - 7%
29757 - 2%
29758 - 11%
29759 - 12%
29760 - 14%
29761 - 0%
29762 - 12%
29763 - 2%
29764 - 2%
29765 - 3%
29766 - 3%
29767 - 3%
29768 - 0%
29769 - 1%
29771 - 0%
29772 - 1%
29773 - 7%
29774 - 12%
29775 - 1%
29776 - 1%
29777 - 1%
29778 - 3%
29779 - 1%
29780 - 1%
29783 - 0%
29784 - 2%
29786 - 0%
29787 - 8%
29788 - 1%
29790 - 1%
29791 - 1%
29792 - 1%
29793 - 10%
29794 - 1%
29795 - 0%
29796 - 1%
29797 - 2%
29798 - 0%
29800 - 2%
29802 - 2%
29803 - 12%
29804 - 7%
29805 - 8%
29806 - 3%
29807 - 0%
29808 - 4%
29809 - 1%
29814 - 1%
29816 - 2%
29817 - 0%
29818 - 1%
29819 - 0%
29821 - 3%
29822 - 8%
29824 - 0%
29825 - 14%
29826 - 0%
29827 - 8%
2982

As we can see here the list is very long so let's have a better view

In [227]:
count=0
for row in df.iloc:
    rows_missing_values = np.mean(row.isnull())
    if round(rows_missing_values*100)>10 :
        print('{} - {}%'.format(row[0], round(rows_missing_values*100)))
        count=count+1
print('{} rows with missing values'.format(count))

2045 - 13%
2266 - 13%
2734 - 11%
2877 - 11%
2900 - 11%
3007 - 12%
3022 - 12%
3159 - 11%
3170 - 11%
3235 - 11%
3257 - 13%
3260 - 13%
3264 - 13%
3269 - 13%
3277 - 13%
3278 - 13%
3279 - 11%
3283 - 13%
3288 - 13%
3289 - 13%
3290 - 13%
3292 - 13%
3296 - 13%
3298 - 13%
3299 - 12%
3300 - 12%
3301 - 11%
3308 - 13%
3309 - 11%
3313 - 12%
3317 - 13%
3320 - 13%
3323 - 11%
3325 - 11%
3328 - 11%
3329 - 14%
3330 - 11%
3335 - 11%
3339 - 12%
3340 - 13%
3345 - 12%
3349 - 11%
3353 - 13%
3357 - 14%
3362 - 13%
3365 - 13%
3368 - 13%
3372 - 13%
3376 - 13%
3384 - 11%
3385 - 13%
3391 - 14%
3400 - 11%
3401 - 11%
3402 - 12%
3407 - 11%
3415 - 12%
3416 - 14%
3417 - 13%
3423 - 11%
3436 - 11%
3440 - 12%
3441 - 13%
3442 - 12%
3443 - 11%
3445 - 11%
3450 - 11%
3451 - 11%
3454 - 11%
3456 - 13%
3459 - 11%
3461 - 13%
3464 - 11%
3465 - 11%
3478 - 11%
3479 - 13%
3481 - 11%
3491 - 13%
3498 - 11%
3516 - 11%
3525 - 11%
3538 - 11%
3539 - 11%
3544 - 11%
3545 - 11%
3552 - 13%
3555 - 11%
3556 - 11%
3561 - 11%
3563 - 11%
3564 - 11%

6521 - 11%
6527 - 12%
6530 - 11%
6540 - 11%
6547 - 12%
6550 - 11%
6551 - 11%
6565 - 12%
6566 - 13%
6568 - 12%
6573 - 12%
6575 - 11%
6583 - 13%
6588 - 12%
6595 - 14%
6600 - 11%
6601 - 12%
6612 - 12%
6619 - 12%
6620 - 12%
6626 - 12%
6630 - 12%
6632 - 12%
6636 - 12%
6639 - 11%
6643 - 12%
6649 - 12%
6655 - 14%
6658 - 11%
6663 - 12%
6667 - 12%
6671 - 11%
6676 - 11%
6680 - 12%
6697 - 11%
6704 - 12%
6714 - 12%
6720 - 11%
6728 - 11%
6738 - 12%
6745 - 11%
6746 - 13%
6749 - 12%
6756 - 14%
6760 - 12%
6771 - 12%
6775 - 14%
6790 - 12%
6793 - 12%
6814 - 12%
6822 - 12%
6823 - 11%
6839 - 12%
6849 - 11%
6852 - 14%
6853 - 12%
6866 - 12%
6873 - 11%
6875 - 11%
6878 - 11%
6884 - 12%
6885 - 12%
6886 - 12%
6888 - 12%
6890 - 12%
6893 - 12%
6894 - 12%
6909 - 12%
6920 - 11%
6923 - 12%
6927 - 14%
6931 - 11%
6941 - 11%
6943 - 12%
6944 - 12%
6945 - 12%
6956 - 12%
6958 - 11%
6966 - 12%
6967 - 12%
6968 - 11%
6969 - 12%
6974 - 12%
6975 - 12%
6980 - 12%
6983 - 12%
6984 - 12%
6987 - 12%
6988 - 12%
6994 - 12%
6999 - 11%

16742 - 12%
16745 - 13%
16751 - 13%
16782 - 12%
16786 - 12%
16834 - 12%
16836 - 13%
16878 - 12%
16882 - 12%
16916 - 12%
16944 - 11%
16953 - 12%
16972 - 13%
17052 - 11%
17053 - 12%
17093 - 13%
17178 - 12%
17189 - 12%
17252 - 13%
17317 - 12%
17365 - 12%
17472 - 12%
17474 - 13%
17505 - 12%
17513 - 13%
17534 - 13%
17569 - 12%
17570 - 12%
17629 - 12%
17638 - 12%
17673 - 12%
17709 - 13%
17711 - 13%
17753 - 12%
17769 - 13%
17789 - 12%
17878 - 13%
17886 - 12%
17898 - 11%
17913 - 12%
17917 - 12%
17921 - 13%
18006 - 12%
18064 - 13%
18163 - 12%
18198 - 13%
18237 - 12%
18252 - 12%
18276 - 13%
18277 - 14%
18283 - 12%
18337 - 13%
18367 - 11%
18371 - 12%
18454 - 12%
18529 - 12%
18543 - 11%
18567 - 12%
18594 - 13%
18614 - 13%
18622 - 12%
18689 - 13%
18717 - 13%
18753 - 11%
18756 - 11%
18783 - 12%
18814 - 12%
18837 - 12%
18872 - 12%
18911 - 12%
18920 - 12%
18963 - 13%
18983 - 12%
18999 - 13%
19003 - 14%
19025 - 13%
19282 - 12%
19314 - 12%
19315 - 12%
19330 - 12%
19333 - 12%
19350 - 12%
19355 - 13%
1939

Now we have only 2206 rows  
So now the question which data should we get rid of?  
I think it's better to get rid of the rows with huge missing values first because row has less impact on data than the column then we reevalute our data

In [ ]:
#print(df.index[df['id']==2045].tolist())


for row in df.iloc:
    rows_missing_values = np.mean(row.isnull())
    if round(rows_missing_values*100)>10 :
        df=df.drop(df.index[df['id']==row[0]].tolist())

print(df.shape)


now let's see the column missing values rate

In [ ]:
count=0
for col in df.columns:
    column_missing_values = np.mean(df[col].isnull())
    if column_missing_values>0 :
        print('{} - {} - {}%'.format(col,df[col].dtypes, round(column_missing_values*100)))
        count=count+1
print('{} columns with missing values'.format(count))

Now as we can see the missing values rate per column has been sgnificatlly decresed but we there still work to be done  
So now will get rid of the columns with huge rate

In [ ]:
print(df.shape)
count=0
for col in df.columns:
    count=count+1
    column_missing_values = np.mean(df[col].isnull())
    if round(column_missing_values*100)>25 :
        del df[col]
        #df=df.drop(columns=count,axis=1)
        count=count-1
print(df.shape)

In this notebook, we are going to replace the not numeric values by the most frequent value and we are going to use the mediane for the numeric values

In [ ]:
#remplcer les valeurs null dans les colones non nummériques par la valeur la plus fréquente

df_categoric = df.select_dtypes(exclude='number')
categoric_cols = df_categoric.columns.values

for col in categoric_cols:
    missing = df[col].isnull()
    num_missing = np.sum(missing)
    
    if num_missing > 0:  # only do the imputation for the columns that have missing values.
        print(col)
        top = df[col].describe()['top'] # impute with the most frequent value.
        
        df[col] = df[col].fillna(top)
        


#remplcer les valeurs null dans les colones nummériques par la valeur la mediane
df_numeric = df.select_dtypes(include=[np.number])

numeric_cols = df_numeric.columns.values

for col in numeric_cols:
    missing = df[col].isnull()
    num_missing = np.sum(missing)
    
    if num_missing > 0:  # only do the imputation for the columns that have missing values.
        med = df[col].median()
        df[col] = df[col].fillna(med)

        df.head()
print(df.shape)

## Traitement des données aberrantes

Now does all the data that we have are correctly placed?  
Unfortantlly no,this can be caused for a lot of reasons for example the user has entred the data wrongly or the data has been worongly mesured.  
So how we are going to identify the wrong data? well in the most cases that data have either a very low value or very high. so let's have a look about our data. 

In [ ]:
for col in df.columns:
    if len(df[col].unique().tolist())<5:
        print(col)
        print(df[col].unique().tolist())

In [ ]:
Q1=df.price_doc.quantile(0.01)
Q3=df.price_doc.quantile(0.99)
ind=df[(df['price_doc'] < Q1 ) | (df['price_doc'] > Q3 )].index
df.drop(ind,axis=0)
print(df.shape)
len(ind)

df_numeric = df.select_dtypes(include=[np.number])
numeric_cols = df_numeric.columns.values

for col in numeric_cols:
    Q1=df.full_sq.quantile(0.25)
    Q3=df.full_sq.quantile(0.75)
    Q=df.full_sq.describe()
    IQR=Q3-Q1
    ind=df[(df[col] < Q1-1.5*IQR ) ].index
    ind2=df[(df[col] > Q3+1.5*IQR ) ].index

    df.loc[ind,col]=Q1
    df.loc[ind2,col]=Q3
print(df.shape)

## Traitement du problème d'incohérence des données 

In this part, we are going to normalize our data for example some strings can have a white spacein the biginnig or in the end thus that make her a whole different string for our model

In [ ]:
df_categoric = df.select_dtypes(exclude='number')
categoric_cols = df_categoric.columns.values

for col in categoric_cols:
    print(col)
    unique=df[col].unique().tolist()
    unique.sort()
    print(unique)

In [ ]:
df_categoric = df.select_dtypes(exclude='number')
categoric_cols = df_categoric.columns.values

for col in categoric_cols:
    #supression des espaces blancs au debut
    df[col] = df[col].str.lstrip()

    #supression des espaces blancs à la fin
    df[col] = df[col].str.rstrip()
    
    #supression des espaces entre les mots
    df[col] = df[col].str.replace(' ','')

    #supression des symboles monétaires
    money_chars = ["$","€","¢","£","₿"]
    df[col]= df[col].replace('money_chars','')

#formater les dates
df['timestamp'] = pd.to_datetime(df['timestamp'],format='%Y-%m-%d')
df.head()
print(df.shape)

## traitement de redondonce des données

In our dataset, all rows must be unique in order to have an accurate resultat so we have to remouve the duplicated rows

In [ ]:
#code ici
#df.drop('price_doc', axis=1).drop_duplicates()
df.drop_duplicates()

print(df.shape)
df.head()


### Transformation des données 

In [ ]:
#code ici
#DFCat=df[categoric_cols]
#DFNum=df[numeric_cols]
#DFCat.head()
#print(DFCat.shape)
#DFCat.head()
#DFCat['culture_objects_top_25']=DFCat['culture_objects_top_25'].replace({"no":0,"yes":1})
#DFCat.replace({"no":0,"yes":1},inplace=True)
from sklearn import preprocessing
#DFCat=df[categoric_cols]
label_encoder = preprocessing.LabelEncoder()
#DFCat['culture_objects_top_25']= label_encoder.fit_transform(DFCat['culture_objects_top_25'])
df=df.replace({"no":0,"yes":1})


df_categoric = df.select_dtypes(exclude='number')
categoric_cols = df_categoric.columns.values

for col in categoric_cols:
    df[col]=label_encoder.fit_transform(df[col])
    

#df = df.drop(columns='product_type')
df.head()

## entrainement des données

In [ ]:
#code ici
from sklearn.model_selection import train_test_split
Data2=df[numeric_cols]
df_y=Data2[['price_doc']]
df_x=Data2.drop(columns=['price_doc'])
print(Data2)

x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size = 0.2, random_state = 4 )

scaler = preprocessing.StandardScaler()
scaler.fit(x_train) 
xtrainnorm=scaler.transform(x_train)           
xtestnorm=scaler.transform(x_test) 
print(df.shape)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

print(xtrainnorm.mean())
print(xtrainnorm.std())

## model excution

Finnally, we are going to excute our model and test his performance.  
In fact,there's a plenty of model types so we are going to test some of them. 

### Linear Regresion


In [ ]:
from sklearn import linear_model

model = linear_model.LinearRegression()

model.fit(x_train,y_train)

results = model.predict(x_test)
print(results)
print(y_test)
#print(y_test)
r2_score = model.score(x_train,y_train)
r3_score = model.score(x_test,y_test)

print(r2_score*100,'%')
print(r3_score*100,'%')

In [ ]:
df.boxplot(column='full_sq')
Q1=df.full_sq.quantile(0.25)
Q3=df.full_sq.quantile(0.75)
Q=df.full_sq.describe()
IQR=Q3-Q1
print(IQR)
print(Q1)
print(Q3)
ind=df[(df['full_sq'] < Q1-1.5*IQR ) ].index
ind2=df[(df['full_sq'] > Q3+1.5*IQR ) ].index

print(df.shape)
print(len(ind2))
print(len(ind))
df.loc[ind,'full_sq']=Q1
df.loc[ind2,'full_sq']=Q3


In [ ]:
from sklearn import preprocessing
Data=df[numeric_cols].drop(columns=['id','price_doc' ],axis=1)

print(df[numeric_cols].shape)
print(Data.shape)

min_max_scaler = preprocessing.MinMaxScaler()
Dataminmax = min_max_scaler.fit_transform(Data) 
mx=df[numeric_cols].max()
mn=df[numeric_cols].min()
print(mx)
print(mn)
print(Dataminmax.max())
print(Dataminmax.min())

In [ ]:
from sklearn.model_selection import train_test_split
Data2=df[numeric_cols]
df_y=Data2[['price_doc']]
df_x=Data2.drop(columns=['price_doc'])

x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size = 0.2, random_state = 4 )

print(x_train.shape)
print(x_test.shape)

scaler = preprocessing.StandardScaler()
scaler.fit(x_train) 
xtrainnorm=scaler.transform(x_train)           
xtestnorm=scaler.transform(x_test)           

print(xtrainnorm.mean())
print(xtrainnorm.std())



In [ ]:
from sklearn import linear_model

model = linear_model.LinearRegression()

model.fit(x_train,y_train)

results = model.predict(x_test)
print(results)
print(y_test)
#print(y_test)
r2_score = model.score(x_train,y_train)
r3_score = model.score(x_test,y_test)

print(r2_score*100,'%')
print(r3_score*100,'%')

In [ ]:
#Utilisation de la règle interquartile pour trouver des valeurs aberrantes
df.boxplot(column='female_f')
Q1=df.female_f.quantile(0.25)
Q3=df.female_f.quantile(0.75)
Q=df.female_f.describe()
IQR=Q3-Q1
print(IQR)
print(Q1)
print(Q3)
ind=df[(df['female_f'] < Q1-1.5*IQR ) ].index
ind2=df[(df['female_f'] > Q3+1.5*IQR ) ].index

print(df.shape)
print(len(ind2))
print(len(ind))
df.loc[ind,'female_f']=Q1
df.loc[ind2,'female_f']=Q3

In [ ]:

from sklearn.model_selection import train_test_split
Data2=df[numeric_cols]
df_y=Data2[['price_doc']]
df_x=Data2.drop(columns=['price_doc'])

x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size = 0.3, random_state = 4 )

print(x_train.shape)
print(x_test.shape)

scaler = preprocessing.StandardScaler()
scaler.fit(x_train) 
xtrainnorm=scaler.transform(x_train)           
xtestnorm=scaler.transform(x_test)           

print(xtrainnorm.mean())
print(xtrainnorm.std())

In [ ]:
from sklearn import linear_model

model = linear_model.LinearRegression()

model.fit(x_train,y_train)

results = model.predict(x_test)
print(results)
print(y_test)
#print(y_test)
r2_score = model.score(x_train,y_train)
r3_score = model.score(x_test,y_test)

print(r2_score*100,'%')
print(r3_score*100,'%')

In [ ]:

df.boxplot(column='office_sqm_500')
Q1=df.office_sqm_500.quantile(0.25)
Q3=df.office_sqm_500.quantile(0.75)
Q=df.office_sqm_500.describe()
IQR=Q3-Q1
print(IQR)
print(Q1)
print(Q3)
ind=df[(df['office_sqm_500'] < Q1-1.5*IQR ) ].index
ind2=df[(df['office_sqm_500'] > Q3+1.5*IQR ) ].index

print(df.shape)
print(len(ind2))
print(len(ind))
df.loc[ind,'office_sqm_500']=Q1
df.loc[ind2,'office_sqm_500']=Q3

In [ ]:
from sklearn.model_selection import train_test_split
Data2=df[numeric_cols]
df_y=Data2[['price_doc']]
df_x=Data2.drop(columns=['price_doc'])

x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size = 0.3, random_state = 8 )

print(x_train.shape)
print(x_test.shape)

scaler = preprocessing.StandardScaler()
scaler.fit(x_train) 
xtrainnorm=scaler.transform(x_train)           
xtestnorm=scaler.transform(x_test)           

print(xtrainnorm.mean())
print(xtrainnorm.std())



In [ ]:
from sklearn import linear_model

model = linear_model.LinearRegression()

model.fit(x_train,y_train)

results = model.predict(x_test)
print(results)
print(y_test)
#print(y_test)
r2_score = model.score(x_train,y_train)
r3_score = model.score(x_test,y_test)

print(r2_score*100,'%')


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

Data2=df[numeric_cols]
df_y=Data2[['price_doc']]
df_x=Data2.drop(columns=['price_doc'])


x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size = 0.2, random_state = 4 )


sc = StandardScaler()
X_train_std = sc.fit_transform(x_train)
X_test_std = sc.transform(x_test)


### KNN models

In [ ]:
from sklearn.preprocessing import KBinsDiscretizer
from sklearn.neighbors import KNeighborsClassifier
kbins = KBinsDiscretizer(n_bins=10, encode='ordinal', strategy='uniform')
ytrain_trans = kbins.fit_transform(y_train)
ytest_trans = kbins.transform(y_test)

model = KNeighborsClassifier(n_neighbors=3)

model.fit(x_train, ytrain_trans)

print(model.score(x_train, ytrain_trans)*100,'%')
print(model.score(x_test, ytest_trans)*100,'%')

### decision tree classifier

In [ ]:
from sklearn.model_selection import RepeatedKFold, RepeatedStratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from numpy import mean
from numpy import std

clf_gini = DecisionTreeClassifier(criterion = "gini", random_state = 100,max_depth=5, min_samples_leaf=20)
  
    # Performing training
cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=3,random_state = 10)
n_scores = cross_val_score(clf_gini, x_train, ytrain_trans, scoring='accuracy', cv=cv, n_jobs=-1, error_score='raise')
# report pipeline performance
print('Accuracy: %.3f (%.3f)' % (mean(n_scores), std(n_scores)))